### ISSUE TITLE
C++: static cast fails to invoke `#convertStatic`

### ISSUE THREAD

**Issue by h0nzZik** [kframework/c-semantics#263]
I have some partial/experimental implementation of  C++ enumerators, and for debugging purpuses I wanted to static_cast them to integers and back. Static cast, as described in 5.2.9:4-5 and implemented in [cpp14/language/translation/expr/cast.k](https://github.com/kframework/c-semantics/blob/master/semantics/cpp14/language/translation/expr/cast.k#L47-L51), first tries to direct-initialize a new temporary variable, and if that fails, it performs some conversion from 5.2.9:6-12 (implemented by `#convertStatic`). However I encountered some problems while casting integers to enums.

Ints are not implicitly convertible to enums, so they should be converted by `#convertStatic`. First problem is that some [rules in cpp14/language/common/conversion.k](https://github.com/kframework/c-semantics/blob/master/semantics/cpp14/language/common/conversion.k#L86-L89) allow implicit conversion of ints to  enums. That is easy to fix.

But after fixing that (and implemeting some productions like `min`, `max` etc) I was not able to make `figureInit()` rewrite to `ill-formed` (or `cannot-convert`) in compile time (while running static semantics). For the cast in this piece of code
```
enum E{};
int main() {
    int a = 4;
    (E)a;
}
```
the translation semantics yields
```
`ExpressionStmt`(lecpp(`ExprLoc`(`CabsLoc`(...),
`_:=__CPP-SYNTAX`(
	lecpp(
		temp(...),
		hasTrace(...),
		tcpp(...,..,unscopedEnum(...))
	),
	`ConvertType`(
		tcpp(...,...,unscopedEnum(...)),
		`ExprLoc`(`CabsLoc`(...),`ExprLoc`(`CabsLoc`(...),
			`Name`(`NoNNS`(.KList),`Identifier`(#token("\"a\"","String")))
		))
	)
)),hasTrace(...),tcpp(...,...,unscopedEnum(...))))
```
which then evaluates to cannot-convert during runtime. I was trying to check the ill-formedness in [assignment.k](https://github.com/kframework/c-semantics/blob/master/semantics/cpp14/language/translation/expr/assignment.k#L10-L12), but without much success, because the `convertType(_,_)` gets rewritten to ConvertType(_,_), which does not get rewritten anymore if the second parameter is a variable.

When trying to convert constant (e.g. `(E)5`), then `convertType` correctly rewrites to `cannot-convert`. @dwightguth any idea?

**Comment by dwightguth** [comment]
so, first of all, you are correct that we should probably add notBool isCPPEnumType(T) to the side condition of the rules in conversion.k. They exist in that form because they need to be able to handle char16_t, char32_t, and wchar_t, which are integer types but neither signed nor unsigned types. It was written without thinking that far ahead about enumerations at the time.

I believe what you want to try to do is to add a side condition to the rule that creates ConvertType nodes in the tree which asserts that the types involved make the conversion possible in some fashion.

**Comment by dwightguth** [comment]
I am working on a fix for this issue and should have it out soon. I would wait and come back to this once that's done.

**Comment by h0nzZik** [comment]
I have already created something for [both](https://github.com/kframework/c-semantics/pull/262/commits/2cbf7f822c9c189c3e1f6245e804c9b10530e33d) [parts](https://github.com/h0nzZik/c-semantics/commit/2958a90882604c25da713d82f921df57206f5538). Do you think it would not be sufficient, or it is too ad-hoc?

**Comment by dwightguth** [comment]
Yes, this is too ad-hoc. I will have something out in the near future that should work without breaking the structure that I already began laying out

**Comment by h0nzZik** [comment]
This was probably fixed in 6cd2af9e88f11039da28a2b7a54bdf5474d0b261. Can we close it?
